### PR TITLE
FMM integration

### PR DESCRIPTION
**Pull request by JoachimMarin**
* Only implemented for LinkedCells
* Contains example and test

# Description

Please include a summary of the changes.

## Related Pull Requests

- #PR

## Resolved Issues

- [ ] #Issue

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: examples/fmm-integration-test/main.cpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#include "../../tests/testAutopas/testingHelpers/RandomGenerator.h"
+
+using AutoPasCont = autopas::AutoPas<autopas::Particle, autopas::FullParticleCell<autopas::Particle>>;
+
+int main(int argc, char **argv) {
+  std::cout << std::endl;
+
+  std::cout << "Init Test" << std::endl;
+
+  auto cont = AutoPasCont();
```
Comment (SteffenSeckler) [inline examples/fmm-integration-test/main.cpp]:
```suggestion
  AutoPasCont cont;
```

File: examples/fmm-integration-test/main.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  cont.setAllowedContainers(std::set<autopas::ContainerOption>{autopas::ContainerOption::linkedCells});
```
Comment (SteffenSeckler) [inline examples/fmm-integration-test/main.cpp]:
```suggestion
  cont.setAllowedContainers({autopas::ContainerOption::linkedCells});
```

File: examples/fmm-integration-test/main.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  cont.setBoxMin(std::array<double, 3>{0, 0, 0});
```
Comment (SteffenSeckler) [inline examples/fmm-integration-test/main.cpp]:
```suggestion
  cont.setBoxMin({0., 0., 0.});
```

File: examples/fmm-integration-test/main.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  cont.setBoxMax(std::array<double, 3>{2, 3, 4});
```
Comment (SteffenSeckler) [inline examples/fmm-integration-test/main.cpp]:
```suggestion
  cont.setBoxMax({2., 3., 4.});
```

File: examples/fmm-integration-test/main.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  cont.init();
+
+  for (int i = 0; i < 10; ++i) {
+    auto r = RandomGenerator::randomPosition(cont.getBoxMin(), cont.getBoxMax());
+    auto v = std::array<double, 3>({0, 0, 0});
+    auto part = autopas::Particle(r, v, i);
+    cont.addParticle(part);
+  }
```
Comment (SteffenSeckler) [inline examples/fmm-integration-test/main.cpp]: use `fillWithParticles()` function of RandomGenerator instead

File: examples/fmm-integration-test/main.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  std::cout << "Number of Particles: " << cont.getNumberOfParticles() << std::endl;
```
Comment (SteffenSeckler) [inline examples/fmm-integration-test/main.cpp]: when you move this to the tests, remove output

File: src/autopas/containers/linkedCells/LinkedCells.h [hunk 7]
```diff
@@ -296,6 +297,100 @@ class LinkedCells : public ParticleContainer<ParticleCell, SoAArraysType> {
...
 
+  void createFmmNode(FmmTreeNode &node, int depth) const {
+    auto nodeMin3DIndex = _cellBlock.get3DIndexOfPosition(node.getBoxMin());
+    auto nodeMax3DIndex = _cellBlock.get3DIndexOfPosition(node.getBoxMax());
+    auto delta = ArrayMath::sub(nodeMax3DIndex, nodeMin3DIndex);
+
+    for (int i = 0; i < depth; ++i) {
+      std::cout << "    ";
+    }
+    std::cout << "createFmmNode" << std::endl;
```
Comment (SteffenSeckler) [inline src/autopas/containers/linkedCells/LinkedCells.h]: use autopas logger instead

File: src/autopas/containers/linkedCells/LinkedCells.h [hunk 8]
```diff
@@ -296,6 +297,100 @@ class LinkedCells : public ParticleContainer<ParticleCell, SoAArraysType> {
...
+      auto midBoxMin = std::array<double, 3>({0, 0, 0});
+      auto midBoxMax = std::array<double, 3>({0, 0, 0});
+      _cellBlock.getCellBoundingBox(split3DIndex, midBoxMin, midBoxMax);
+
+      // std::cout << "midBox" << std::endl;
+
+      // std::cout << autopas::ArrayUtils::to_string(midBoxMin) << std::endl;
+      // std::cout << autopas::ArrayUtils::to_string(midBoxMax) << std::endl;
+
+      auto splitMin = std::array<double, 3>({0, 0, 0});
```
Comment (SteffenSeckler) [inline src/autopas/containers/linkedCells/LinkedCells.h]:
```suggestion
      auto splitMin = node.getBoxMin();
```

File: src/autopas/containers/linkedCells/LinkedCells.h [hunk 9]
```diff
@@ -296,6 +297,100 @@ class LinkedCells : public ParticleContainer<ParticleCell, SoAArraysType> {
...
+      auto splitMax = std::array<double, 3>({0, 0, 0});
```
Comment (SteffenSeckler) [inline src/autopas/containers/linkedCells/LinkedCells.h]:
```suggestion
      auto splitMax = node.getBoxMax();
```

File: src/autopas/fastMultipoleMethod/FmmTreeNode.h [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+#include <vector>
+
+class FmmTree;
+
+class FmmTreeNode {
+ public:
+  FmmTreeNode(FmmTree &tree, FmmTreeNode *parent, std::array<double, 3> boxMin, std::array<double, 3> boxMax);
+  void split(std::array<double, 3> firstBoxMax, std::array<double, 3> secondBoxMin);
+  FmmTreeNode &getChild(int index) { return *child[index]; }
+  [[nodiscard]] const std::array<double, 3> &getBoxMin() const { return boxMin; }
```
Comment (SteffenSeckler) [inline src/autopas/fastMultipoleMethod/FmmTreeNode.h]:
```suggestion
  [[nodiscard]] std::array<double, 3> getBoxMin() const { return boxMin; }
```

File: src/autopas/fastMultipoleMethod/FmmTreeNode.h [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  [[nodiscard]] const std::array<double, 3> &getBoxMax() const { return boxMax; }
```
Comment (SteffenSeckler) [inline src/autopas/fastMultipoleMethod/FmmTreeNode.h]:
```suggestion
  [[nodiscard]] std::array<double, 3> getBoxMax() const { return boxMax; }
```

File: src/autopas/containers/linkedCells/LinkedCells.h [hunk 12]
```diff
@@ -296,6 +297,100 @@ class LinkedCells : public ParticleContainer<ParticleCell, SoAArraysType> {
...
+
+      for (int i = 0; i < 3; ++i) {
```
Comment (SteffenSeckler) [inline src/autopas/containers/linkedCells/LinkedCells.h]: remove loop

**Comment by SteffenSeckler** [comment]
inside of autopas:
* please always use the `autopas` namespace.
* use additional namespace for fmm (`fmm`)

File: src/autopas/fastMultipoleMethod/FmmTreeNode.h [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+      interactionList.insert(&otherNode);
+    } else {
+      auto boxSize1 = utils::ArrayMath::sub(boxMax, boxMin);
+      auto boxSize2 = utils::ArrayMath::sub(otherNode.boxMax, otherNode.boxMin);
+
+      auto largestIndex1 = findLargestIndex(boxSize1);
+      auto largestIndex2 = findLargestIndex(boxSize2);
+
+      auto largestSize1 = boxSize1[largestIndex1];
+      auto largestSize2 = boxSize2[largestIndex2];
```
Comment (SteffenSeckler) [inline src/autopas/fastMultipoleMethod/FmmTreeNode.h]: you should probably move this into a member function, e.g, `getLargestSize()` as it makes the code a bit easier to read.

**Comment by FG-TUM** [comment]
Closed because FMM integration is not easily doable. See #4 for discussion.

**Pull request closed without merging**